pumpkin-solver/src/propagators/arithmetic/linear_less_or_equal.rs

```diff
@@ -41,57 +74,97 @@ where
-        "LinearLeq"
+    fn notify(
+        &mut self,
+        context: PropagationContext,
+        local_id: LocalId,
+        _event: OpaqueDomainEvent,
+    ) -> EnqueueDecision {
+        let index = local_id.unpack() as usize;
+
+        let x_i = &self.x[index];
+        let old_bound = self.current_bounds[index];
+        let new_bound = context.lower_bound(x_i);
+
+        pumpkin_assert_simple!(
+            old_bound < new_bound,
+            "propagator should only be triggered when lower bounds are tightened, old_bound={old_bound}, new_bound={new_bound}"
+        );
+
+        self.current_bounds[index] = new_bound;
+        self.lower_bound_left_hand_side += new_bound - old_bound;
+
+        EnqueueDecision::Enqueue
     }
 
-    fn debug_propagate_from_scratch(&self, context: PropagationContextMut) -> PropagationStatusCP {
-        perform_propagation(context, &self.x, self.c)
+    fn synchronise(&mut self, context: PropagationContext) {
+        self.recalculate_incremental_state(context);
     }
-}
 
-fn perform_propagation<Var: IntegerVariable>(
-    mut context: PropagationContextMut,
-    x: &[Var],
-    c: i32,
-) -> PropagationStatusCP {
-    let lb_lhs = x.iter().map(|var| context.lower_bound(var)).sum::<i32>();
-    if c < lb_lhs {
-        let reason: PropositionalConjunction = x
-            .iter()
-            .map(|var| predicate![var >= context.lower_bound(var)])
-            .collect();
-        return Err(reason.into());
+    fn priority(&self) -> u32 {
+        0
     }
 
-    for (i, x_i) in x.iter().enumerate() {
-        let bound = c - (lb_lhs - context.lower_bound(x_i));
+    fn name(&self) -> &str {
+        "LinearLeq"
+    }
 
-        if context.upper_bound(x_i) > bound {
-            let reason: PropositionalConjunction = x
-                .iter()
-                .enumerate()
-                .filter_map(|(j, x_j)| {
-                    if j != i {
-                        Some(predicate![x_j >= context.lower_bound(x_j)])
-                    } else {
-                        None
-                    }
-                })
-                .collect();
+    fn debug_propagate_from_scratch(
+        &self,
+        mut context: PropagationContextMut,
+    ) -> PropagationStatusCP {
+        if let Some(conjunction) = self.detect_inconsistency(context.as_readonly()) {
+            return Err(conjunction.into());
+        }
 
-            context.set_upper_bound(x_i, bound, reason)?;
+        for (i, x_i) in self.x.iter().enumerate() {
+            let bound = self.c - (self.lower_bound_left_hand_side - context.lower_bound(x_i));
+
+            if context.upper_bound(x_i) > bound {
+                let reason: PropositionalConjunction = self
+                    .x
+                    .iter()
+                    .enumerate()
+                    .filter_map(|(j, x_j)| {
+                        if j != i {
+                            Some(predicate![x_j >= context.lower_bound(x_j)])
```

Comment:
Maybe a bit out of the scope of the PR but would it make sense to filter out the elements which do not contribute to the `lower_bound_left_hand_side` (i.e. the tasks which have a lower-bound of 0)?